From b64ec77f73397b175b207799140b6106ce7b5b7a Mon Sep 17 00:00:00 2001
From: Shaun Tancheff <shaun@tancheff.com>
Date: Wed, 9 Mar 2016 10:44:53 -0600
Subject: [PATCH 8/8] Stream Id: Use PID to seed Stream Id construction.

Use PID to differentiate between streams of incoming
data to simulate the effect that stream id may have
on certain predictable workloads.

Signed-off-by: Shaun Tancheff <shaun.tancheff@seagate.com>
---
 block/bio.c               |  2 ++
 drivers/md/dm-zoned.c     | 20 +++++++++++++++++---
 fs/ext4/ext4.h            |  1 +
 fs/ext4/file.c            |  1 +
 fs/ext4/inode.c           |  4 ++++
 fs/ext4/page-io.c         |  8 +++++++-
 include/linux/blk_types.h |  7 +++++++
 include/linux/fs.h        |  6 ++++++
 8 files changed, 45 insertions(+), 4 deletions(-)

diff --git a/block/bio.c b/block/bio.c
index a874abb..541ba10 100644
--- a/block/bio.c
+++ b/block/bio.c
@@ -582,6 +582,7 @@ void __bio_clone_fast(struct bio *bio, struct bio *bio_src)
 	bio->bi_rw = bio_src->bi_rw;
 	bio->bi_iter = bio_src->bi_iter;
 	bio->bi_io_vec = bio_src->bi_io_vec;
+	bio->pid = bio_src->pid;
 }
 EXPORT_SYMBOL(__bio_clone_fast);
 
@@ -664,6 +665,7 @@ struct bio *bio_clone_bioset(struct bio *bio_src, gfp_t gfp_mask,
 	bio->bi_rw		= bio_src->bi_rw;
 	bio->bi_iter.bi_sector	= bio_src->bi_iter.bi_sector;
 	bio->bi_iter.bi_size	= bio_src->bi_iter.bi_size;
+	bio->pid		= bio_src->pid;
 
 	if (bio->bi_rw & REQ_DISCARD)
 		goto integrity_clone;
diff --git a/drivers/md/dm-zoned.c b/drivers/md/dm-zoned.c
index ade1cb7..db561de 100644
--- a/drivers/md/dm-zoned.c
+++ b/drivers/md/dm-zoned.c
@@ -102,12 +102,26 @@ static void zdm_remove_proc_entries(struct zoned *znd);
  */
 static inline u32 bio_stream(struct bio *bio)
 {
+	u32 stream_id = 0x40;
+
 	/*
 	 * Since adding stream id to a BIO is not yet in mainline we just
-	 * assign some defaults: use stream_id 0xff for upper level meta data
-	 * and 0x40 for everything else ...
+	 * use this heuristic to try to skip unnecessary co-mingling of data.
 	 */
-	return (bio->bi_rw & REQ_META) ? 0xff : 0x40;
+
+	if (bio->bi_rw & REQ_META)
+		stream_id = 0xff;
+	else {
+		u32 upid = bio->pid;
+
+		stream_id = ((upid/97) + (upid/1031) + (upid)) & 0xff;
+		if (stream_id == 0)
+			stream_id++;
+		if (stream_id == 0xff)
+			stream_id--;
+	}
+
+	return stream_id;
 }
 
 /**
diff --git a/fs/ext4/ext4.h b/fs/ext4/ext4.h
index f5e9f04..8733231 100644
--- a/fs/ext4/ext4.h
+++ b/fs/ext4/ext4.h
@@ -191,6 +191,7 @@ struct ext4_io_submit {
 	struct bio		*io_bio;
 	ext4_io_end_t		*io_end;
 	sector_t		io_next_block;
+	pid_t			pid;
 };
 
 /*
diff --git a/fs/ext4/file.c b/fs/ext4/file.c
index bc313ac..a8e95c3 100644
--- a/fs/ext4/file.c
+++ b/fs/ext4/file.c
@@ -260,6 +260,7 @@ static int ext4_file_open(struct inode * inode, struct file * filp)
 	char buf[64], *cp;
 	int ret;
 
+	inode->pid = get_current()->pid;
 	if (unlikely(!(sbi->s_mount_flags & EXT4_MF_MNTDIR_SAMPLED) &&
 		     !(sb->s_flags & MS_RDONLY))) {
 		sbi->s_mount_flags |= EXT4_MF_MNTDIR_SAMPLED;
diff --git a/fs/ext4/inode.c b/fs/ext4/inode.c
index cecf9aa..1192063 100644
--- a/fs/ext4/inode.c
+++ b/fs/ext4/inode.c
@@ -1840,6 +1840,7 @@ static int ext4_writepage(struct page *page,
 		return __ext4_journalled_writepage(page, len);
 
 	ext4_io_submit_init(&io_submit, wbc);
+	io_submit.pid = inode->pid;
 	io_submit.io_end = ext4_init_io_end(inode, GFP_NOFS);
 	if (!io_submit.io_end) {
 		redirty_page_for_writepage(wbc, page);
@@ -2465,6 +2466,7 @@ static int ext4_writepages(struct address_space *mapping,
 	mpd.inode = inode;
 	mpd.wbc = wbc;
 	ext4_io_submit_init(&mpd.io_submit, wbc);
+	mpd.io_submit.pid = inode->pid;
 retry:
 	if (wbc->sync_mode == WB_SYNC_ALL || wbc->tagged_writepages)
 		tag_pages_for_writeback(mapping, mpd.first_page, mpd.last_page);
@@ -2518,6 +2520,8 @@ retry:
 			}
 		}
 		ext4_journal_stop(handle);
+		if (mpd.io_submit.io_bio)
+			mpd.io_submit.io_bio->pid = inode->pid;
 		/* Submit prepared bio */
 		ext4_io_submit(&mpd.io_submit);
 		/* Unlock pages we didn't use */
diff --git a/fs/ext4/page-io.c b/fs/ext4/page-io.c
index 5602450..3f396f5 100644
--- a/fs/ext4/page-io.c
+++ b/fs/ext4/page-io.c
@@ -358,6 +358,7 @@ void ext4_io_submit(struct ext4_io_submit *io)
 
 	if (bio) {
 		bio_get(io->io_bio);
+		io->io_bio->pid = io->pid;
 		submit_bio(io->io_op, io->io_bio);
 		bio_put(io->io_bio);
 	}
@@ -386,6 +387,7 @@ static int io_submit_init_bio(struct ext4_io_submit *io,
 	bio->bi_end_io = ext4_end_bio;
 	bio->bi_private = ext4_get_io_end(io->io_end);
 	io->io_bio = bio;
+	io->io_bio->pid = io->pid;
 	io->io_next_block = bh->b_blocknr;
 	return 0;
 }
@@ -468,8 +470,10 @@ int ext4_bio_write_page(struct ext4_io_submit *io,
 			/* A hole? We can safely clear the dirty bit */
 			if (!buffer_mapped(bh))
 				clear_buffer_dirty(bh);
-			if (io->io_bio)
+			if (io->io_bio) {
+				io->io_bio->pid = inode->pid;
 				ext4_io_submit(io);
+			}
 			continue;
 		}
 		if (buffer_new(bh)) {
@@ -494,6 +498,8 @@ int ext4_bio_write_page(struct ext4_io_submit *io,
 	do {
 		if (!buffer_async_write(bh))
 			continue;
+		if (io->io_bio)
+			io->io_bio->pid = inode->pid;
 		ret = io_submit_add_bh(io, inode,
 				       data_page ? data_page : page, bh);
 		if (ret) {
diff --git a/include/linux/blk_types.h b/include/linux/blk_types.h
index 7303b34..e42ceac 100644
--- a/include/linux/blk_types.h
+++ b/include/linux/blk_types.h
@@ -70,6 +70,7 @@ struct bio {
 	bio_end_io_t		*bi_end_io;
 
 	void			*bi_private;
+
 #ifdef CONFIG_BLK_CGROUP
 	/*
 	 * Optional ioc and css associated with this bio.  Put on bio
@@ -84,6 +85,12 @@ struct bio {
 #endif
 	};
 
+/*
+ * SST: HACK: most recent process id to open a file is held via BIO
+ *      for later use as *stream_id*.
+ */
+	pid_t			pid;
+
 	unsigned short		bi_vcnt;	/* how many bio_vec's */
 
 	/*
diff --git a/include/linux/fs.h b/include/linux/fs.h
index 84b783f..a5fc67b 100644
--- a/include/linux/fs.h
+++ b/include/linux/fs.h
@@ -676,6 +676,12 @@ struct inode {
 	struct hlist_head	i_fsnotify_marks;
 #endif
 
+/*
+ * SST: HACK most recent process id to open a file is held
+ *      via inode for passing to BIO.
+ */
+	pid_t			pid;
+
 	void			*i_private; /* fs or device private pointer */
 };
 
-- 
2.7.0

